### PR TITLE
Add folder pattern for AddOp and MulOp

### PR DESCRIPTION
**Pull request by qingyunqu**
Add folder pattern:
* mhlo.add(%0, 0) -> %0
* mhlo.mul(%0, 1) -> %0

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -4390,27 +4390,67 @@ struct min<APInt> {
...
 BINARY_FOLDER(DivOp, divide);
 BINARY_FOLDER(RemOp, remainder);
 BINARY_FOLDER(MaxOp, max);
 BINARY_FOLDER(MinOp, min);
 
+OpFoldResult AddOp::fold(ArrayRef<Attribute> attrs) {
+  if (attrs[0] && attrs[1]) {
+    BINARY_FOLDER_INTERNAL(AddOp, std::plus)
+  }
+  if (attrs[0] || attrs[1]) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4419]: Can you add a comment about what will be handled?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 2]
```diff
@@ -4390,27 +4390,67 @@ struct min<APInt> {
...
+    }
+  }
+  return {};
+}
+
+OpFoldResult MulOp::fold(ArrayRef<Attribute> attrs) {
+  if (attrs[0] && attrs[1]) {
+    BINARY_FOLDER_INTERNAL(MulOp, std::multiplies);
+  }
+  if (attrs[0] || attrs[1]) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4438]: (same here)

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 1 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4419]:
```suggestion
  // Handle special case where one operand is 0:  x + 0 => x
  if (attrs[0] || attrs[1]) {
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4439]:
```suggestion
  // Handle special case where one operand is 1: x * 1 => x
  if (attrs[0] || attrs[1]) {
```

**Review by joker-eph — APPROVED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 1 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4420]: Thanks for adding comments !

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 3]
```diff
@@ -4390,27 +4390,69 @@ struct min<APInt> {
...
+  }
+  // Handle special case where one operand is 0:  x + 0 => x
+  if (attrs[0] || attrs[1]) {
+    SplatElementsAttr attr = attrs[0] ? attrs[0].dyn_cast<SplatElementsAttr>()
+                                      : attrs[1].dyn_cast<SplatElementsAttr>();
+    if (!attr) return {};
+    Value result = attrs[0] ? rhs() : lhs();
+    if (attr.getElementType().isa<FloatType>()) {
+      if (attr.getSplatValue<APFloat>().isZero()) return result;
+
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4427]: Nit: could you delete this empty lline?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 4]
```diff
@@ -4390,27 +4390,69 @@ struct min<APInt> {
...
+  }
+  // Handle special case where one operand is 1: x * 1 => x
+  if (attrs[0] || attrs[1]) {
+    SplatElementsAttr attr = attrs[0] ? attrs[0].dyn_cast<SplatElementsAttr>()
+                                      : attrs[1].dyn_cast<SplatElementsAttr>();
+    if (!attr) return {};
+    Value result = attrs[0] ? rhs() : lhs();
+    if (attr.getElementType().isa<FloatType>()) {
+      if (attr.getSplatValue<APFloat>().convertToDouble() == 1.0) return result;
+
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:4447]: Same

**Review by jpienaar — APPROVED** [review]
I was going to ask about the equivalent of https://llvm.discourse.group/t/canonicalization-of-x-0-0-in-tosa/4760 , but I see for HLOs we don't consider this.

**Comment by qingyunqu** [comment]
> I was going to ask about the equivalent of https://llvm.discourse.group/t/canonicalization-of-x-0-0-in-tosa/4760 , but I see for HLOs we don't consider this.

There are some models lowered from tf dialect or torch_xla, they are not lowered from tosa. So I think these folders are meaningful.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 3 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: Done, thanks.

**Review by joker-eph — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**